File: python/ch1/ex108.py
import unittest
import math


def cbrt(x):
    return iter(1.0, x)


def iter(guess, x):
    return guess if is_good_enough(guess, x) else iter(improve(guess, x), x)


def improve(guess, x):
    return (x / square(guess) + 2 * guess) / 3.0


def square(x):
    return x * x


def is_good_enough(guess, x):
    return abs(cube(guess) - x) < (guess / 100000.0)


def cube(x):
    return x * x * x


# ------------------------------

class MyTest(unittest.TestCase):

    def test_root(self):
        self.assertAlmostEqual(cube(cbrt(27.0)), 27.0, places=4)
        self.assertAlmostEqual(cube(cbrt(7.0)), 7.0, places=4)
        self.assertAlmostEqual(cube(cbrt(0.027)), 0.027, places=4)



if __name__ == '__main__':
    unittest.main()

File: java/src/ch1/ex108.java
package ch1;

import java.util.function.Function;

public class ex108 {
    static class CubeRootFunction {
        static final double factor = 100000.0;

        double calc(double x) {
            return iter(1.0, x);
        }

        double iter(double guess, double x) {
            return is_good_enough(guess, x) ? guess : iter(improve(guess, x), x);
        }

        boolean is_good_enough(double guess, double x) {
            return Math.abs(cube(guess) - x) < (guess / factor);
        }

        double cube(double x) {
            return x * x * x;
        }

        double improve(double guess, double x) {
            return (2 * guess + x / (guess * guess)) / 3;
        }

        double average(double x, double y) {
            return (x + y) / 2.0;
        }
    }

    static class cbrt extends CubeRootFunction implements Function<Double, Double> {
        @Override
        public Double apply(Double x) {
            return calc(x);
        }
    }

    public static void main(String[] args) {
        final cbrt f = new cbrt();
        System.out.println(f.apply(64.0) + " vs. cubed\n" + f.cube(f.apply(64.0)));
        System.out.println(f.apply(27.0) + " vs. cubed\n" + f.cube(f.apply(27.0)));
        System.out.println(f.apply(0.027) + " vs. cubed\n" + f.cube(f.apply(0.027)));
    }
}
